src/plugins/data/common/search/aggs/buckets/lib/histogram_calculate_interval.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { isAutoInterval } from '../_interval_options';
+
+interface IntervalValuesRange {
+  min: number;
+  max: number;
+}
+
+export interface CalculateHistogramIntervalParams {
+  interval: string;
+  maxBucketsUiSettings: number;
+  maxBucketsUserInput?: number;
+  intervalBase?: number;
+  values?: IntervalValuesRange;
+}
+
+/**
+ * Round interval by order of magnitude to provide clean intervals
+ */
+const roundInterval = (minInterval: number) => {
+  const orderOfMagnitude = Math.pow(10, Math.floor(Math.log10(minInterval)));
+  let interval = orderOfMagnitude;
+
+  while (interval < minInterval) {
+    interval += orderOfMagnitude;
+  }
+
+  return interval;
+};
+
+const calculateForGivenInterval = (
+  diff: number,
+  interval: number,
+  maxBucketsUiSettings: CalculateHistogramIntervalParams['maxBucketsUiSettings']
+) => {
+  const bars = diff / interval;
+
+  if (bars > maxBucketsUiSettings) {
+    const minInterval = diff / maxBucketsUiSettings;
+
+    return roundInterval(minInterval);
+  }
+
+  return interval;
+};
+
+const calculateAutoInterval = (
```

Comment:
These functions are not behaving the way I expected, as described here: https://github.com/elastic/kibana/issues/75438#issuecomment-676571091

I'll comment separately with tweaks.